ambry-frontend/src/main/java/com/github/ambry/frontend/NamedBlobPutHandler.java

```diff
@@ -548,5 +559,67 @@ private void addDatasetVersion(BlobProperties blobProperties, RestRequest restRe
             RestServiceErrorCode.getRestServiceErrorCode(ex.getErrorCode()));
       }
     }
+
+    /**
+     * Support delete the dataset version out of retentionCount
+     */
+    private void deleteDatasetVersionOutOfRetentionCount() {
+      Dataset dataset = (Dataset) restRequest.getArgs().get(RestUtils.InternalKeys.TARGET_DATASET);
+      try {
+        String accountName = dataset.getAccountName();
+        String containerName = dataset.getContainerName();
+        String datasetName = dataset.getDatasetName();
+        List<DatasetVersionRecord> datasetVersionRecordList =
+            accountService.getAllValidVersionsOutOfRetentionCount(accountName, containerName, datasetName);
+        if (datasetVersionRecordList.size() > 0) {
+          DatasetVersionRecord record = datasetVersionRecordList.get(0);
+          String version = record.getVersion();
+          RequestPath requestPath = getRequestPath(restRequest);
+          RequestPath newRequestPath =
+              new RequestPath(requestPath.getPrefix(), requestPath.getClusterName(), requestPath.getPathAfterPrefixes(),
+                  NAMED_BLOB_PREFIX + SLASH + accountName + SLASH + containerName + SLASH + datasetName + SLASH
+                      + version, requestPath.getSubResource(), requestPath.getBlobSegmentIdx());
+          // Replace RequestPath in the RestRequest and call DeleteBlobHandler.handle.
+          restRequest.setArg(InternalKeys.REQUEST_PATH, newRequestPath);
+          restRequest.setArg(InternalKeys.TARGET_ACCOUNT_KEY, null);
+          restRequest.setArg(InternalKeys.TARGET_CONTAINER_KEY, null);
+          deleteBlobHandler.handle(restRequest, restResponseChannel,
+              recursiveCallback(datasetVersionRecordList, 1, accountName, containerName, datasetName));
+        }
+      } catch (Exception e) {
+        LOGGER.error("Failed to delete dataset version out of retention count for this dataset: " + dataset.toString());
+      }
+    }
+
+    /**
+     * Recursively delete the dataset version out of retention count at dataset level.
+     * @param datasetVersionRecordList the list of the {@link DatasetVersionRecord}
+     * @param idx the index of the datasetVersionRecordList
+     * @param accountName the name of the account.
+     * @param containerName the name of the container.
+     * @param datasetName the name of the dataset.
+     */
+    private Callback<Void> recursiveCallback(List<DatasetVersionRecord> datasetVersionRecordList, int idx, String accountName,
+        String containerName, String datasetName) {
+      if (idx == datasetVersionRecordList.size()) {
```

Comment:
Can we add some print message(or metric) to record the status of the background deletion?

Reply:
Sure, I'll add it in a follow up pr.